builder/amazon/common/ssm_driver.go

```diff
@@ -46,20 +53,51 @@ func NewSSMDriver(config SSMDriverConfig) *SSMDriver {
 // not wish to manage the session manually calling StopSession on a instance of this driver will terminate the active session
 // created from calling StartSession.
 func (d *SSMDriver) StartSession(ctx context.Context, input ssm.StartSessionInput) (*ssm.StartSessionOutput, error) {
+	d.wg.Add(1)
+	defer d.wg.Done()
 	log.Printf("Starting PortForwarding session to instance %q", aws.StringValue(input.Target))
 
-	var output *ssm.StartSessionOutput
-	err := retry.Config{
-		ShouldRetry: func(err error) bool { return IsAWSErr(err, "TargetNotConnected", "") },
-		RetryDelay:  (&retry.Backoff{InitialBackoff: 200 * time.Millisecond, MaxBackoff: 60 * time.Second, Multiplier: 2}).Linear,
-	}.Run(ctx, func(ctx context.Context) (err error) {
-		output, err = d.SvcClient.StartSessionWithContext(ctx, &input)
-		return err
-	})
+	output, err := d.StartSessionWithContext(ctx, input)
 	if err != nil {
 		return nil, fmt.Errorf("error encountered in starting session for instance %q: %s", aws.StringValue(input.Target), err)
 	}
 
+	d.retryConnection = make(chan bool, 1)
+	d.retryAfterTermination = make(chan bool, 1)
+	d.closeRetryPolling = make(chan bool, 1)
+	// Starts go routine that will keep listening to channels and retry the session creation/connection when needed.
+	// The log polling process will add data to the channels whenever a retryable error happens to the session or if it's terminated.
+	go func(ctx context.Context, driver *SSMDriver, input ssm.StartSessionInput) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case ok := <-d.closeRetryPolling:
+				if ok {
+					return
+				}
+			case r := <-driver.retryAfterTermination:
```

Comment:
Here we can check if the channel is close and return if it is. The same can be done for the retryConnection channel and at that point I believe we can get rid of the closeRetryPolling channel.